keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+reduces the opportunities for runtime errors.
+
+Also, making it possible (and convenient) to declare "zero trust" policies is
+important to compliance. By "zero trust", we mean policy rules that apply the
+principle of least privilege to newly created resources (e.g. namespaces) where
+the policy is initially set to the most restrictive state and can be made less
+restrictive via configuration.
+
+## Proposal
+
+Introduce a new `ValidatingAdmissionPolicy` kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to the match rules of admission webhooks, RBAC, priority & fairness and Audit)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  but with access to the data in
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Version conversion support (similar to admission webhook's MatchPolicy)
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+failure policies, type safety, advanced matching rules, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "replicalimit-policy.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  match:
+    - resourceRules:
+      - apiGroups:   ["apps"]
+        apiVersions: ["v1"]
+        operations:  ["CREATE", "UPDATE"]
+        resources:   ["deployments"]
+  validations:
+    - expression: "object.spec.replicas <= config.maxReplicas"
+      reason: Invalid
+      messageExpression: "'object.spec.replicas must be no greater than ' + config.maxReplicas"
```

Comment:
nit related to [this comment](https://github.com/kubernetes/enhancements/pull/3492/files#r977164813) - maybe this should be `messageExpression: "'object.spec.replicas must be no greater than ' + string(config.maxReplicas)"` instead? 
I tried making it a CRD validation rule - `size('object.spec.replicas must be no greater than ' + self.maxReplicas) < 75` so it'd evaluate to a bool - and got `ERROR: <input>:1:54: found no matching overload for '_+_' applied to '(string, int)'`.

Reply:
Good find. I'll update the example. This does make message formatting a bit less pleasant, which makes me wonder if we'd be better off just offering a simple template format like "object.spec.replicas must be no greater than {{config.maxReplicas}}" where only dereferencing is supported.

Reply:
Were you thinking of using [text/template](https://pkg.go.dev/text/template)? The syntax reminded me of it, and since it's in the standard library I don't think we'd need to do much work to add it in. It's rather complex/Turing-complete (would be kinda funny to have error message templates be Turing-complete while the actual policy rules weren't, heh), so maybe using it would be a bit too much.

We could roll our own with a similar syntax, but there's a couple things to think about:
* Array subscripting - would be simple for users (`{{ config.someArray[2] }}`) but we'd have to parse it/write support for it.
* Type checking - we'd have to implement some kind of system even if we want just informational type checking. As an aside, as far as I know `text/template` lacks a type system for what that's worth.

We could also try for a simple string-replace system, though I don't see how we could do subscripts/type-checking that way, so I don't really think I'm a fan of that route, at least on paper.

I agree that having to potentially write `string(someVar)` would make CEL-based message expressions a little unwieldly, though at the same time I'm a little concerned about the potential complexity other options would involve.

Reply:
I'm comfortable ruling out text/template. If we go for simpler string formatting, json path is probably the least risky choice since CRDs already use them for a few things. Right now I'm tempted to just stick with CEL.

@TristonianJones any recommendations for string formatting with CEL?

Reply:
Oh, one other thought here, we could do either:

`object.spec.replicas must be no greater than {{<cel expression>}}`

Or,

```yaml
messageFormat: "object.spec.replicas must be no greater than {0}"
messageArgs: ["<cel expression>"]
```

This eliminates the need to do the type casts to string.

Reply:
I like embedding CEL expressions with `{{ }}`. Since it also prevents the need to nest quotes (`" ' ' "`), maybe we could end up rolling the `messageExpression` field into `message`, since for error messages without CEL expressions they'd end up being equivalent?

Reply:
Embedding `{{<cel expression}}` in strings does mean that all CEL expressions must have escaping rules defined to handle `{{` or `}}` inside the CEL expression. If we use `message` and `messageArgs` we don't have to deal with this, but it makes the simple cases a bit more verbose to write.

Reply:
@jpbetz @DangerOnTheRanger Is the primary issue that making `message` a CEL expression would require type-casting of parameters? There's always the option of introducing a `string.format("message {1}, {2}", [args])` method into CEL. I'm not sure you need to invent a new syntax around it in the YAML.